sql/catalog/lease: TestDescriptorRefreshOnRetry failed

sql/catalog/lease.TestDescriptorRefreshOnRetry [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/18100453?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/18100453?buildTab=artifacts#/) on release-23.2 @ [b14741589afbd8876aa6818bc0bbe07ecc4da017](https://github.com/cockroachdb/cockroach/commits/b14741589afbd8876aa6818bc0bbe07ecc4da017):


```
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:163 +0x22
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x56
        created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 201018
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x96
        
        goroutine 204774 [select]:
        github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1({0xc015aed800, 0xc01344dc01})
        	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:156 +0xee
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:479 +0x13a
        created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 204744
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x415
        
        goroutine 199747 [select]:
        github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1({0xc00a9f2c00, 0x0})
        	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:156 +0xee
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:479 +0x13a
        created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 199718
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x415
        
        goroutine 200610 [select]:
        github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).processEvents(0xc005a3ba20, {0x62966d8, 0xc00f32aff0}, 0xc011dc86c0?, 0xc003f7dec0)
        	github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed/rangefeed.go:351 +0xd4
        github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run.func2({0x62966d8?, 0xc00f32aff0?})
        	github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed/rangefeed.go:308 +0x32
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:163 +0x22
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x56
        created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 199943
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x96
        
        goroutine 205048 [select]:
        github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).processEvents(0xc0183cedc0, {0x62966d8, 0xc018af6a00}, 0xc012b33110?, 0xc0183dea20)
        	github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed/rangefeed.go:351 +0xd4
        github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run.func2({0x62966d8?, 0xc018af6a00?})
        	github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed/rangefeed.go:308 +0x32
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:163 +0x22
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x56
        created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 204964
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x96
        
    lease_test.go:970: invalid descriptor acquisition counts = 2, 0
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/ce8d73b4a4081f39d36a7fa2c2508fcc/logTestDescriptorRefreshOnRetry13277273
--- FAIL: TestDescriptorRefreshOnRetry (2.48s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDescriptorRefreshOnRetry.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-45366

--- a/pkg/sql/catalog/lease/BUILD.bazel
+++ b/pkg/sql/catalog/lease/BUILD.bazel
@@ -134,6 +134,7 @@ go_test(
         "//pkg/sql/sqlliveness",
         "//pkg/sql/sqlliveness/slbase",
         "//pkg/sql/sqlliveness/slprovider",
+        "//pkg/sql/stats",
         "//pkg/sql/types",
         "//pkg/storage",
         "//pkg/testutils",

--- a/pkg/sql/catalog/lease/lease_test.go
+++ b/pkg/sql/catalog/lease/lease_test.go
@@ -56,6 +56,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slbase"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slprovider"
+	"github.com/cockroachdb/cockroach/pkg/sql/stats"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
@@ -949,14 +950,16 @@ func TestDescriptorRefreshOnRetry(t *testing.T) {
 			},
 		},
 	}
+	params.Settings = cluster.MakeTestingClusterSettings()
+	// Disable the automatic stats collection, which could interfere with
+	// the lease acquisition counts in this test.
+	stats.AutomaticStatisticsClusterMode.Override(ctx, &params.Settings.SV, false)
+	// Set a long lease duration so that the periodic task to refresh leases does
+	// not run.
+	lease.LeaseDuration.Override(ctx, &params.Settings.SV, 24*time.Hour)
 	srv, sqlDB, kvDB := serverutils.StartServer(t, params)
 	defer srv.Stopper().Stop(context.Background())
 	s := srv.ApplicationLayer()
-	// Disable the automatic stats collection, which could interfere with
-	// the lease acquisition counts in this test.
-	if _, err := sqlDB.Exec("SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false"); err != nil {
-		t.Fatal(err)
-	}
 	if _, err := sqlDB.Exec(`
 CREATE DATABASE t;
 CREATE TABLE t.foo (v INT);

